PR#16 ( renderNewsletterSignupCallout )

rgadon107 opened the pull request:
## PR Summary 

This pull request includes unit and integration tests for the function `render_newsletter_signup_callout()` in the `extend-give-wp` plugin. The relative file path to the function under test is:

>`/extend-give-wp/src/support/load-assets.php`.

rgadon107 committed:
Commit unit and integration test files.

hellofromtonya commented on tests/phpunit/Unit/support/renderNewsletterSignupCallout.php:
```diff
@@ -0,0 +1,77 @@
+			->with( 'form_id' )
+			->andReturn( $post_data['form_id'] );
+		Functions\expect( '_get_plugin_dir' )->andReturn( EXTEND_GIVE_WP_ROOT_DIR );
+		Functions\when( 'esc_attr' )->justEcho( $post_data['form_id'] );
```
`esc_attr() does not echo; rather it returns.

Check out how we do it here. You can add this to your test case to stub a bunch of different functions https://github.com/wp-media/phpunit/blob/master/Unit/TestCase.php#L63

hellofromtonya replied:
Just checked. [You have that method in the test case](https://github.com/rgadon107/extend-give-wp/blob/master/tests/phpunit/Unit/TestCase.php#L46). You have to call it in order for it work though. You can do that in the `setUp()` method.

```php
	public function setUp() {
		parent::setUp();

		$this->setup_common_wp_stubs();
	}
```

rgadon107 replied:
@hellofromtonya As I wrote earlier, i have no idea why I overlooked your comment from last evening. It didn't show up on my mental radar until late this afternoon. 

I added the method to the `setUp()` method of the unit test for `renderFormFeaturedImageAndCaption.php`, and resolved the error with `esc_attr()` being undefined. 

I'm curious though; why not add that method to the `setUp()` method for the unit TestCase instead?

hellofromtonya commented on tests/phpunit/Unit/support/renderNewsletterSignupCallout.php:
```diff
@@ -0,0 +1,77 @@
+	public function addTestData() {
+		return [
+			'newsletter callout view' => [
+				'post_data'     => [
```
Will there be other post data? If no, your test is only using the form ID. Instead, just pass that ID instead of the associative array. Less code. Easier to understand.

rgadon107 replied:
@hellofromtonya Good point. I updated the dataProvider accordingly.

hellofromtonya commented on tests/phpunit/Unit/support/renderNewsletterSignupCallout.php:
```diff
@@ -0,0 +1,77 @@
+				,
+			],
+		];
+	}
```
Add more test data scenarios such as no form ID.

rgadon107 replied:
@hellofromtonya Great call!

When I added a data set for an empty form_id and expected_view, I realized I need to add a guardrail conditional to return early if the the value of `$form_id` in the function under test is empty. This affects 4 of the 6 callbacks registered to action hooks in `/extend-give-wp/src/support/load-assets.php`.  I will update both the source code and test branches accordingly.

rgadon107 committed:
Update return method for mock of 'esc_attr'.

- The function 'esc_attr' returns the string parameter. It does not echo.

rgadon107 committed:
Add data set to test that $form_id & $expected_view are empty.

rgadon107 committed:
Add 'setup_common_wp_stubs' from unit TestCase to 'setUp()' method.

rgadon107 committed:
Merge branch 'master' into renderNewsletterSignupCallout

rgadon107 committed:
Update integration test file. 

- Change the dataProvider params for the test method.
- Relocate the $post_id factory() method to the dataProvider. 
- Add 2 data set; 'test data is empty', & 'test data is valid'.
- Fix expected HTML markup to match actual value.

rgadon107 committed:
phpcs fix.
